File: apps/api/src/tradingos_api/schemas/morning_plan.py
from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel

from tradingos_api.models.enums import (
    AlertDeliveryStatus,
    DeliveryChannel,
    MorningPlanSectionKey,
    MorningPlanVersionLabel,
    PlanCompletenessStatus,
)


class MorningPlanQualityCheckResponse(BaseModel):
    check_name: str
    passed: bool
    detail: str | None

    model_config = {"from_attributes": True}


class MorningPlanItemResponse(BaseModel):
    id: uuid.UUID
    recommendation_version_id: uuid.UUID | None
    instrument_id: uuid.UUID | None = None
    display_order: int
    headline: str
    action_label: str | None = None
    # "Every card must expose evidence, deterministic calculations, AI
    # synthesis, policy result, and user/broker state separately" — the
    # five keys below are `card_detail`'s fixed shape
    # (`services/morning_plan_generate.py`); returned as one JSON object
    # rather than five separate response fields purely to avoid
    # duplicating the shape in two places, not because the separation
    # is any less real.
    card_detail: dict[str, Any] = {}

    model_config = {"from_attributes": True}


class TopStatusResponse(BaseModel):
    market_date: date
    is_trading_day: bool
    market_closed_reason: str | None
    countdown_to_open_seconds: int | None
    plan_status: str
    plan_version_id: uuid.UUID | None
    plan_version_label: MorningPlanVersionLabel | None
    generated_at: datetime | None
    evidence_cutoff: datetime | None
    provider_broker_status: str
    regime_classification: str | None
    vix_proxy_level: Decimal | None
    vix_percentile: Decimal | None
    total_equity: Decimal
    cash: Decimal
    exposure_pct: Decimal
    risk_budget_pct: Decimal | None
    operating_mode: str
    kill_switch_active: bool

    model_config = {"from_attributes": True}


class MorningPlanSectionResponse(BaseModel):
    section_key: MorningPlanSectionKey
    display_order: int
    items: list[MorningPlanItemResponse]


class MorningPlanDeliveryEventResponse(BaseModel):
    channel: DeliveryChannel
    status: AlertDeliveryStatus
    delivered_at: datetime | None

    model_config = {"from_attributes": True}


class MorningPlanVersionSummaryResponse(BaseModel):
    id: uuid.UUID
    morning_plan_run_id: uuid.UUID
    plan_date: date
    version_label: MorningPlanVersionLabel
    version_number: int
    evidence_cutoff: datetime
    generated_at: datetime
    completeness_status: PlanCompletenessStatus

    model_config = {"from_attributes": True}


class MorningPlanVersionDetailResponse(MorningPlanVersionSummaryResponse):
    sections: list[MorningPlanSectionResponse]
    quality_checks: list[MorningPlanQualityCheckResponse]
    delivery_events: list[MorningPlanDeliveryEventResponse]


class DashboardResponse(BaseModel):
    """The Morning Decision Dashboard — top status plus the fixed
    section hierarchy. `version` is `None` exactly when `top_status.plan_status`
    is `MARKET_CLOSED`, `INCOMPLETE` (nothing generated yet), or `FAILED`
    — a dashboard with no plan is a legitimate, honestly-labeled state,
    never backfilled with a fabricated one."""

    top_status: TopStatusResponse
    version: MorningPlanVersionDetailResponse | None


class GeneratePlanRequest(BaseModel):
    plan_date: date | None = None  # defaults to today in America/Los_Angeles
    version_label: MorningPlanVersionLabel = MorningPlanVersionLabel.AD_HOC
    triggered_by: str = "manual"
    idempotency_key: str | None = None

File: apps/web/lib/api/morningPlan.ts
import { apiGet } from "./client";

export type MorningPlanSectionKey =
  | "ACT_NOW"
  | "APPROVAL_REQUIRED"
  | "HOLD_MANAGE"
  | "INVESTMENT_WATCH"
  | "TACTICAL_WATCH"
  | "AVOID"
  | "DATA_PROBLEMS"
  | "BUY_AND_HOLD"
  | "TACTICAL_TRADES"
  | "WATCH_AND_AVOID"
  | "UPCOMING_EVENTS";

export type MorningPlanVersionLabel = "PRELIMINARY" | "FINAL" | "AD_HOC" | "CORRECTION";
export type PlanCompletenessStatus = "COMPLETE" | "INCOMPLETE" | "NO_ACTION" | "FAILED";

/**
 * `card_detail`'s five fixed keys (docs/MORNING_PLAN_SPEC.md /
 * services/morning_plan_generate.py): evidence, deterministic
 * calculations, AI synthesis, policy result, and user/broker state —
 * always present, sometimes empty objects. Never merged into one blob
 * client-side; each drill-in surface reads its own key so the
 * evidence-vs-inference distinction the backend deliberately keeps
 * separate stays visible in the UI too.
 */
export type MorningPlanCardDetail = {
  evidence: Record<string, unknown>;
  deterministic: Record<string, unknown>;
  ai_synthesis: Record<string, unknown>;
  policy_result: Record<string, unknown>;
  user_broker_state: Record<string, unknown>;
};

export type MorningPlanItem = {
  id: string;
  recommendation_version_id: string | null;
  instrument_id: string | null;
  display_order: number;
  headline: string;
  action_label: string | null;
  card_detail: MorningPlanCardDetail;
};

export type MorningPlanSection = {
  section_key: MorningPlanSectionKey;
  display_order: number;
  items: MorningPlanItem[];
};

export type MorningPlanQualityCheck = {
  check_name: string;
  passed: boolean;
  detail: string | null;
};

export type MorningPlanDeliveryEvent = {
  channel: "IN_APP" | "COWORK" | "MARKDOWN_EXPORT";
  status: string;
  delivered_at: string | null;
};

export type MorningPlanVersionDetail = {
  id: string;
  morning_plan_run_id: string;
  plan_date: string;
  version_label: MorningPlanVersionLabel;
  version_number: number;
  evidence_cutoff: string;
  generated_at: string;
  completeness_status: PlanCompletenessStatus;
  sections: MorningPlanSection[];
  quality_checks: MorningPlanQualityCheck[];
  delivery_events: MorningPlanDeliveryEvent[];
};

export type TopStatus = {
  market_date: string;
  is_trading_day: boolean;
  market_closed_reason: string | null;
  countdown_to_open_seconds: number | null;
  plan_status: string;
  plan_version_id: string | null;
  plan_version_label: MorningPlanVersionLabel | null;
  generated_at: string | null;
  evidence_cutoff: string | null;
  provider_broker_status: string;
  regime_classification: string | null;
  vix_proxy_level: string | null;
  vix_percentile: string | null;
  total_equity: string;
  cash: string;
  exposure_pct: string;
  risk_budget_pct: string | null;
  operating_mode: string;
  kill_switch_active: boolean;
};

export type DashboardResponse = {
  top_status: TopStatus;
  version: MorningPlanVersionDetail | null;
};

export function getMorningPlanDashboard(params?: {
  planDate?: string;
  now?: string;
}): Promise<DashboardResponse> {
  const query = new URLSearchParams();
  if (params?.planDate) query.set("plan_date", params.planDate);
  if (params?.now) query.set("now", params.now);
  const qs = query.toString();
  return apiGet<DashboardResponse>(`/api/v1/morning-plan/dashboard${qs ? `?${qs}` : ""}`);
}
